### PR TITLE
1158 - edge_native_base

### PR DESCRIPTION
**Pull request by jychuah**
Closes [#1158](https://github.com/enthought/edge-product/issues/1158)

## Changes

- .dev folder contains edge_native_base and edge_hello_world
- edm environment names match name of application
- CI uses config.py to set variables to reduce copy/paste configuration in CI
- CI inserts "NATIVE_APP_MODE" environment variable
- CI adds commands for specifying edge environment variables (not fully functional until release of Edge 2.5.2)
- edge-native-base configures nginx to run as app user
- edge-native-base has separate app.location.conf.template that only exposes client application nginx configuration
- edge-hello-world uses app user to install enthought_edge from bundle and  run install and run application
- edge-hello-world reports activity

## Testing

- In `.dev/edge-native-base` run `python bootstrap.py`
- `edm shell -e edge-native-base`
- `python -m ci build`
- In `.dev/edge-hello-world` run `python bootstrap.py`
- `edm shell -e edge-hello-world`
- `python -m ci generate_bundle`, `python -m ci build`
- `python -m ci standalone` - Hello World should be browseable at http://localhost:8888
- `python -m ci start` - Hello World should be launchable from local JupyterHub at http://localhost:8000

### CODE REVIEW

File: .dev/edge-hello-world/ci/__main__.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    env["IMAGE_NAME"] = IMAGE_NAME
+    env["IMAGE_TAG"] = tag
+    subprocess.run(cmd, check=True, env=env)
+    click.echo("JupyterHub is running at: http://127.0.0.1:8888")
+
+
+@cli.command("standalone")
+@click.option("--tag", default=IMAGE_TAG, help="Docker tag to use.")
+@click.pass_obj
+def start(obj, tag):
```
Comment (andrewcollette) [inline .dev/edge-hello-world/ci/__main__.py:121]: should use a different name for the function

File: .dev/edge-native-base/README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
+# Edge OAuth2 Native Base Image 
+
+This folder contains the source for building the `quay.io/enthought/edge-oauth2-app` Docker
```
Comment (andrewcollette) [inline .dev/edge-native-base/README.md:3]: name is stale (but see above, let's nuke most of the readme)

File: .dev/edge-native-base/bootstrap.py [hunk 3]
```diff
@@ -22,7 +22,6 @@
 PIP_DEPS = [
     "jupyterhub==2.2.2",
     "sqlalchemy<2",
-    "configurable-http-proxy",
```
Comment (andrewcollette) [inline .dev/edge-native-base/bootstrap.py:25]: ✅

File: .dev/edge-native-base/ci/__main__.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    "packaging",
+    "pip",
+    "pyparsing",
+    "setuptools",
+    "six",
+    "click",
+]
+
+@click.group()
+@click.option("--edge-api-url", default="https://edge.enthought.com/services/api", help="The API url for Edge")
```
Comment (andrewcollette) [inline .dev/edge-native-base/ci/__main__.py:36]: Do we need all of these for the base image build?

**Review by andrewcollette — APPROVED** [review]
Couple minor comments.  Also, for some reason doing `python -m ci start` for the hello-world example launched an old version of the streamlit demo 🤷‍♂️ .  Maybe something I have to clear out of Docker?

**Comment by jychuah** [comment]
> Couple minor comments. Also, for some reason doing `python -m ci start` for the hello-world example launched an old version of the streamlit demo 🤷‍♂️ . Maybe something I have to clear out of Docker?

This needs to be addressed in https://github.com/enthought/edge-product/issues/1160. This is a rough edge when it comes to JupyterHub:

- If "remove container" is set to True, then this problem will not occur but container logs will not be available if the application fails during preflight due to removal of the container
- If "remove container" is set to False, then this problem will occur if the previous container (which has the inconvenient name of `jupyter-<username>`) is not pruned prior to launch, resulting in a stale container

Likely, the solution will have to be one of the following:

- set the "remove container" setting to False and `docker container prune` before preflight launch, which impacts the developer's local system by pruning all containers (the specific container cannot be targeted for pruning without knowing the username from jupyterhub from a previous run)
- set the "remove container" flag to True and document that the developer must check container logs during a failed preflight validation *without* shutting down jupyterhub (assuming that the spawner does not attempt to remove the failed container immediately)
- override the Dockerspawner to remove any existing stale containers before attempting to launch

**Pull request merged**